### PR TITLE
Fix issue 399 - Remove processing of linear geometries for speedup

### LINKED ISSUE #399: Remove processing of linear geometries for speedup
The curved geometry evaluation processes all geometry types, even if they are linear or planar. It would bring performance benefits to just omit these geometries

### PR DESCRIPTION
**Pull request by jfussbro**
**_Describe your changes here:_**

Before this PR all geometries were linked to occ geometries (when using curved option) even if they are linear.
This PR skips linear geometries when linking to occ geometries, resulting in a slightly better performance. There is now the posibility to use curved meshes in the time_forest_partition script.

Closes #399 

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

- [ ] The author added a BSD statement to `doc/` (or already has one)
- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)
- [ ] New source/header files are properly added to the Makefiles
- [ ] New Datatypes are added to t8indent_custom_datatypes.txt
- [ ] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [ ] New tests use the Google Test framework
- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

### CODE REVIEW

File: benchmarks/time_forest_partition.cxx [hunk 1]
```diff
@@ -327,8 +328,15 @@ t8_time_forest_create_cmesh (const char *msh_file, int mesh_dim,
   if (msh_file != NULL) {
     /* Create a cmesh from the given mesh files */
     cmesh =
-      t8_cmesh_from_msh_file ((char *) msh_file, 0, comm, mesh_dim, 0, 0);
-    partition = 1;
+      t8_cmesh_from_msh_file ((char *) msh_file, 0, comm, mesh_dim, 0, use_occ);
+    // partition = 1;
```
Comment (sandro-elsweijer) [inline benchmarks/time_forest_partition.cxx:332]: Please remove this code fragment

File: benchmarks/time_forest_partition.cxx [hunk 2]
```diff
@@ -327,8 +328,15 @@ t8_time_forest_create_cmesh (const char *msh_file, int mesh_dim,
...
+    if (use_occ) {
+      partition = 0;
+      t8_global_productionf("There will be no partition of the mesh due to the use of the curved mesh option. "
```
Comment (sandro-elsweijer) [inline benchmarks/time_forest_partition.cxx:335]:
```suggestion
      t8_global_productionf("The cmesh is not partitioned due to the usage of the curved mesh option. "
```

File: benchmarks/time_forest_partition.cxx [hunk 3]
```diff
@@ -327,8 +328,15 @@ t8_time_forest_create_cmesh (const char *msh_file, int mesh_dim,
...
+                            "Timing will not be comparable to non-curved meshes.");
+    } else {
+      partition = 0;
```
Comment (sandro-elsweijer) [inline benchmarks/time_forest_partition.cxx:338]:
```suggestion
      partition = 1;
```

File: benchmarks/time_forest_partition.cxx [hunk 4]
```diff
@@ -337,8 +345,10 @@ t8_time_forest_create_cmesh (const char *msh_file, int mesh_dim,
      * among all ranks */
     cmesh = t8_cmesh_load_and_distribute (cmesh_file, num_files, comm,
                                           T8_LOAD_STRIDE, stride);
-    /* Partition only if more than 1 input file */
-    partition = num_files > 1;
+    /* Partition only if more than 1 input file and no curved option*/
+    if (num_files > 1 && !use_occ) {
+      partition = num_files > 1;
+    }
```
Comment (sandro-elsweijer) [inline benchmarks/time_forest_partition.cxx:351]: This bit of code is only executed, when no `msh` file is specified, and we only support curved geometries for `msh` files. Therefore, this change here is not necessary. But we should abort, if the user uses the `use_occ` option but does not provide an `msh` file along with it

File: benchmarks/time_forest_partition.cxx [hunk 5]
```diff
@@ -450,6 +460,8 @@ main (int argc, char *argv[])
                          "Create ghost elements.");
   sc_options_add_switch (opt, 'b', "balance", &do_balance,
                          "Establish a 2:1 balance in the forest.");
+  sc_options_add_switch (opt, 'z', "use_occ", &use_occ,
+                         "If used, meshes will be curved to original geometries (msh- and brep-files nessesary).");
```
Comment (sandro-elsweijer) [inline benchmarks/time_forest_partition.cxx:464]:
```suggestion
                         "If used, meshes will be curved to original geometries (msh- and brep-files necessary).");
```

File: src/t8_cmesh/t8_cmesh_readmshfile.cxx [hunk 6]
```diff
@@ -1429,6 +1434,10 @@ t8_cmesh_msh_file_4_read_eles (t8_cmesh_t cmesh, FILE *fp,
                   edge_nodes[i_edge_node].entity_dim = 2;
                 }
               }
+              if (occ_geometry->t8_geom_is_plane(edge_geometry_tag))
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_readmshfile.cxx:1437]:
```suggestion
              if (occ_geometry->t8_geom_is_line(edge_geometry_tag))
```

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.hxx [hunk 7]
```diff
@@ -259,6 +259,20 @@ public:
                                             const double edge_param, 
                                             double* face_params) const;
 
+  /** Check if an edge is linear or not.
+   * \param [in]  edge_index     The index of the occ edge.
+   * \return                     0 if edge is line.
```
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.hxx:264]:
```suggestion
   * \return                     1 if edge is line.
```

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.hxx [hunk 8]
```diff
@@ -259,6 +259,20 @@ public:
...
+   */
+  int
+  t8_geom_is_line(const int edge_index) const;
+
+  /** Check if a face is plane or not.
+   * \param [in]  face_index     The index of the occ face.
+   * \return                     0 if face is plane.
```
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.hxx:271]:
```suggestion
   * \return                     1 if face is plane.
```

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: benchmarks/time_forest_partition.cxx [hunk 9]
```diff
@@ -339,6 +346,8 @@ t8_time_forest_create_cmesh (const char *msh_file, int mesh_dim,
                                           T8_LOAD_STRIDE, stride);
     /* Partition only if more than 1 input file */
     partition = num_files > 1;
+    /* Abort if use_occ option selected but no msh file provided */
```
Comment (sandro-elsweijer) [inline benchmarks/time_forest_partition.cxx:349]: This check should be placed [here](https://github.com/DLR-AMR/t8code/blob/3f73e1715f9a3c76ed7a8c03f0e00a14bdcfb846/benchmarks/time_forest_partition.cxx#L469-L477). The checks for wron user input are always handled directly after the parsing of the CLI options.

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]
I have one last request and then we can merge

File: benchmarks/time_forest_partition.cxx [hunk 10]
```diff
@@ -463,7 +472,7 @@ main (int argc, char *argv[])
       || T <= 0 || test_tet + test_linear_cylinder + test_occ_cylinder > 1
       || (cmesh_level >= 0 && (!test_linear_cylinder && !test_occ_cylinder))
       || ((mshfileprefix != NULL || cmeshfileprefix != NULL)
-          && (test_linear_cylinder || test_occ_cylinder || test_tet))) {
+          && (test_linear_cylinder || test_occ_cylinder || test_tet || use_occ))) {
```
Comment (sandro-elsweijer) [inline benchmarks/time_forest_partition.cxx:475]: This change results in not being able to specify a `msh` file and using the `use_occ` option at the same time. The program is only executed, if the `if` statement is false

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

**Review by chiaraMaHe — APPROVED** [review]

**Comment by sandro-elsweijer** [comment]
I am afraid the reformatting deleted all the changes you made @jfussbro. So the skipping of linear geometries is completely missing now.

**Review by sandro-elsweijer — APPROVED** [review]

**Pull request merged**